pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ReaderTest.java

```diff
@@ -397,4 +398,70 @@ public void testKeyHashRangeReader() throws IOException {
         }
 
     }
+
+    @Test
+    public void testReaderSubName() throws Exception {
+        doTestReaderSubName(true);
+        doTestReaderSubName(false);
+    }
+
+    public void doTestReaderSubName(boolean setPrefix) throws Exception {
+        final String topic = "persistent://my-property/my-ns/testReaderSubName" + System.currentTimeMillis();
+        final String subName = "my-sub-name";
+
+        ReaderBuilder<String> builder = pulsarClient.newReader(Schema.STRING)
+                .subscriptionName(subName)
+                .topic(topic)
+                .startMessageId(MessageId.earliest);
+        if (setPrefix) {
+            builder = builder.subscriptionRolePrefix(subName + System.currentTimeMillis());
+        }
+        Reader<String> reader = builder.create();
+        ReaderImpl<String> readerImpl = (ReaderImpl<String>) reader;
+        assertEquals(readerImpl.getConsumer().getSubscription(), subName);
+        reader.close();
+
+        final String topic2 = "persistent://my-property/my-ns/testReaderSubName2" + System.currentTimeMillis();
+        admin.topics().createPartitionedTopic(topic2, 3);
+        builder = pulsarClient.newReader(Schema.STRING)
+                .subscriptionName(subName)
+                .topic(topic2)
+                .startMessageId(MessageId.earliest);
+        if (setPrefix) {
+            builder = builder.subscriptionRolePrefix(subName + System.currentTimeMillis());
+        }
+        reader = builder.create();
+        MultiTopicsReaderImpl<String> multiTopicsReader = (MultiTopicsReaderImpl<String>) reader;
+        multiTopicsReader.getMultiTopicsConsumer().getConsumers()
+                .forEach(consumerImpl -> assertEquals(consumerImpl.getSubscription(), subName));
+        multiTopicsReader.close();
+    }
+
+    @Test
+    public void testSameSubName() throws Exception {
+        final String topic = "persistent://my-property/my-ns/testSameSubName";
+        final String subName = "my-sub-name";
+
+        Reader<String> reader = pulsarClient.newReader(Schema.STRING)
+                .subscriptionName(subName)
+                .topic(topic)
+                .startMessageId(MessageId.earliest).create();
+        Reader<String> reader2 = null;
+        try {
+            reader2 = pulsarClient.newReader(Schema.STRING)
+                    .subscriptionName(subName)
+                    .topic(topic)
+                    .startMessageId(MessageId.earliest).create();
+            fail("should fail");
+        } catch (PulsarClientException e) {
+            assertTrue(e instanceof PulsarClientException.ConsumerBusyException);
+            assertTrue(e.getMessage().contains("Exclusive consumer is already connected"));
+        }
+
+        reader.close();
+        if (reader2 != null) {
+            reader2.close();
+        }
+    }
```

Comment:
Can we assert that after closing the first reader it is possible to create a new reader with the same subscription name?